In [1]:
import pandas as pd # A general purpose Python library for data analysis
import numpy as np # A library for scientific computing in Python (e.g., provides high-performance multi-dimensional array objects and operations)

import matplotlib.pyplot as plt # a plotting library for Python and NumPy (readily customizable)
import seaborn as sns # Another plotting library for Python (fewer syntax, excellent default themes, behind the scenes, it uses matplotlib)
import time

## Knowledge Stream Summer 2023

In this notebook, we will learn about the key data structures provided by the Pandas library: **Data Frames, Series, and Indices**.

In addition, we will learn about the following operations:
* How to access data contained in these structures?
* How to read files (e.g., csv, xlsx, sql) to create these structures?
* How to carry out different data manipulation tasks using these structures?

`Dataset`: US elections with information about candidates, their party, votes won, year of election and the result.

## Reading in Data Frames from Files

Pandas has a number of useful file reading tools. You can see them enumerated by typing **"pd.re"** and pressing `tab`. We'll be using **read_csv** today. Note that these file reading functions do all the *data parsing* for you, which is very useful.

Before loading a file into a dataframe, let's first take a look at the **elections.csv** file

In [3]:
#Answer Here

x = pd.read_csv('/workspaces/Advance_DS/elections.csv')
print('\n\n.CSV File Data That read : \n\n', x)




.CSV File Data That read : 

    Candidate        Party     %  Year Result
0     Reagan   Republican  50.7  1980    win
1     Carter   Democratic  41.0  1980   loss
2   Anderson  Independent   6.6  1980   loss
3     Reagan   Republican  58.8  1984    win
4    Mondale   Democratic  37.6  1984   loss
5       Bush   Republican  53.4  1988    win
6    Dukakis   Democratic  45.6  1988   loss
7    Clinton   Democratic  43.0  1992    win
8       Bush   Republican  37.4  1992   loss
9      Perot  Independent  18.9  1992   loss
10   Clinton   Democratic  49.2  1996    win
11      Dole   Republican  40.7  1996   loss
12     Perot  Independent   8.4  1996   loss
13      Gore   Democratic  48.4  2000   loss
14      Bush   Republican  47.9  2000    win
15     Kerry   Democratic  48.3  2004   loss
16      Bush   Republican  50.7  2004    win
17     Obama   Democratic  52.9  2008    win
18    McCain   Republican  45.7  2008   loss
19     Obama   Democratic  51.1  2012    win
20    Romney   Republic

We can use the **head command** to show only a few rows of a dataframe.

In [4]:
x = pd.read_csv('/workspaces/Advance_DS/elections.csv', index_col='Candidate')
print(x)

                 Party     %  Year Result
Candidate                                
Reagan      Republican  50.7  1980    win
Carter      Democratic  41.0  1980   loss
Anderson   Independent   6.6  1980   loss
Reagan      Republican  58.8  1984    win
Mondale     Democratic  37.6  1984   loss
Bush        Republican  53.4  1988    win
Dukakis     Democratic  45.6  1988   loss
Clinton     Democratic  43.0  1992    win
Bush        Republican  37.4  1992   loss
Perot      Independent  18.9  1992   loss
Clinton     Democratic  49.2  1996    win
Dole        Republican  40.7  1996   loss
Perot      Independent   8.4  1996   loss
Gore        Democratic  48.4  2000   loss
Bush        Republican  47.9  2000    win
Kerry       Democratic  48.3  2004   loss
Bush        Republican  50.7  2004    win
Obama       Democratic  52.9  2008    win
McCain      Republican  45.7  2008   loss
Obama       Democratic  51.1  2012    win
Romney      Republican  47.2  2012   loss
Clinton     Democratic  48.2  2016

In [5]:
a = x.reset_index()
print(a)

   Candidate        Party     %  Year Result
0     Reagan   Republican  50.7  1980    win
1     Carter   Democratic  41.0  1980   loss
2   Anderson  Independent   6.6  1980   loss
3     Reagan   Republican  58.8  1984    win
4    Mondale   Democratic  37.6  1984   loss
5       Bush   Republican  53.4  1988    win
6    Dukakis   Democratic  45.6  1988   loss
7    Clinton   Democratic  43.0  1992    win
8       Bush   Republican  37.4  1992   loss
9      Perot  Independent  18.9  1992   loss
10   Clinton   Democratic  49.2  1996    win
11      Dole   Republican  40.7  1996   loss
12     Perot  Independent   8.4  1996   loss
13      Gore   Democratic  48.4  2000   loss
14      Bush   Republican  47.9  2000    win
15     Kerry   Democratic  48.3  2004   loss
16      Bush   Republican  50.7  2004    win
17     Obama   Democratic  52.9  2008    win
18    McCain   Republican  45.7  2008   loss
19     Obama   Democratic  51.1  2012    win
20    Romney   Republican  47.2  2012   loss
21   Clint

# heading
## heading2

In [6]:
# Answer Here
b = a.head()
print('By default first five rows show :\n' , b)

By default first five rows show :
   Candidate        Party     %  Year Result
0    Reagan   Republican  50.7  1980    win
1    Carter   Democratic  41.0  1980   loss
2  Anderson  Independent   6.6  1980   loss
3    Reagan   Republican  58.8  1984    win
4   Mondale   Democratic  37.6  1984   loss


In [7]:
b = a.head(10)
print('First ten (10) rows show :\n' , b)

First ten (10) rows show :
   Candidate        Party     %  Year Result
0    Reagan   Republican  50.7  1980    win
1    Carter   Democratic  41.0  1980   loss
2  Anderson  Independent   6.6  1980   loss
3    Reagan   Republican  58.8  1984    win
4   Mondale   Democratic  37.6  1984   loss
5      Bush   Republican  53.4  1988    win
6   Dukakis   Democratic  45.6  1988   loss
7   Clinton   Democratic  43.0  1992    win
8      Bush   Republican  37.4  1992   loss
9     Perot  Independent  18.9  1992   loss


There is also a **tail command**.

In [ ]:
#Answer Here
c = a.tail()
print('Last five rows show \n' , c)

Last five rows show 
    Candidate       Party     %  Year Result
18    McCain  Republican  45.7  2008   loss
19     Obama  Democratic  51.1  2012    win
20    Romney  Republican  47.2  2012   loss
21   Clinton  Democratic  48.2  2016   loss
22     Trump  Republican  46.1  2016    win


The `read_csv` command lets us specify a **column to use an index**. For example, we could have used __Year__ as the index.

In [ ]:
#Answer Here
d = a.set_index('Year')
print('Now Year as an index :\n' , d)


Now Year as an index :
      Candidate        Party     % Result
Year                                    
1980    Reagan   Republican  50.7    win
1980    Carter   Democratic  41.0   loss
1980  Anderson  Independent   6.6   loss
1984    Reagan   Republican  58.8    win
1984   Mondale   Democratic  37.6   loss
1988      Bush   Republican  53.4    win
1988   Dukakis   Democratic  45.6   loss
1992   Clinton   Democratic  43.0    win
1992      Bush   Republican  37.4   loss
1992     Perot  Independent  18.9   loss
1996   Clinton   Democratic  49.2    win
1996      Dole   Republican  40.7   loss
1996     Perot  Independent   8.4   loss
2000      Gore   Democratic  48.4   loss
2000      Bush   Republican  47.9    win
2004     Kerry   Democratic  48.3   loss
2004      Bush   Republican  50.7    win
2008     Obama   Democratic  52.9    win
2008    McCain   Republican  45.7   loss
2012     Obama   Democratic  51.1    win
2012    Romney   Republican  47.2   loss
2016   Clinton   Democratic  48.2

Alternately, we could have used the **set_index** commmand on the dataframe.

In [ ]:
#Answer Here
# As we read file already
dataFrame = pd.DataFrame(a.set_index('Year'))
print('Now Year as an index using DataFrame:\n' , dataFrame)
dataFrame.reset_index()

Now Year as an index using DataFrame:
               Candidate                  Party  Popular vote Result          %
Year                                                                          
1824     Andrew Jackson  Democratic-Republican        151271   loss  57.210122
1824  John Quincy Adams  Democratic-Republican        113142    win  42.789878
1828     Andrew Jackson             Democratic        642806    win  56.203927
1828  John Quincy Adams    National Republican        500897   loss  43.796073
1832     Andrew Jackson             Democratic        702735    win  54.574789
...                 ...                    ...           ...    ...        ...
2016         Jill Stein                  Green       1457226   loss   1.073699
2020       Joseph Biden             Democratic      81268924    win  51.311515
2020       Donald Trump             Republican      74216154   loss  46.858542
2020       Jo Jorgensen            Libertarian       1865724   loss   1.177979
2020     Howa

Year          Candidate                  Party  Popular vote Result  \
0    1824     Andrew Jackson  Democratic-Republican        151271   loss   
1    1824  John Quincy Adams  Democratic-Republican        113142    win   
2    1828     Andrew Jackson             Democratic        642806    win   
3    1828  John Quincy Adams    National Republican        500897   loss   
4    1832     Andrew Jackson             Democratic        702735    win   
..    ...                ...                    ...           ...    ...   
177  2016         Jill Stein                  Green       1457226   loss   
178  2020       Joseph Biden             Democratic      81268924    win   
179  2020       Donald Trump             Republican      74216154   loss   
180  2020       Jo Jorgensen            Libertarian       1865724   loss   
181  2020     Howard Hawkins                  Green        405035   loss   

             %  
0    57.210122  
1    42.789878  
2    56.203927  
3    43.796073  
4    54.574789  
..         ...  
177   1.073699  
178  51.311515  
179  46.858542  
180   1.177979  
181   0.255731  

[182 rows x 6 columns]

In [9]:
# Here we read .csv file directly

dataFrame_1 = pd.DataFrame((pd.read_csv('/workspaces/Advance_DS/elections.csv', index_col='Year')))
print('DataFrame using Their on DataFrame Structure\n\n',dataFrame_1)
dataFrame_1.reset_index()
print('\n\n')
dataFrame_2 = pd.DataFrame((pd.read_csv('/workspaces/Advance_DS/elections.csv').set_index('Year')))
print('\n\nDataFrame without Their on DataFrame Structure\n\n',dataFrame_2)
dataFrame_2.reset_index()

DataFrame using Their on DataFrame Structure

      Candidate        Party     % Result
Year                                    
1980    Reagan   Republican  50.7    win
1980    Carter   Democratic  41.0   loss
1980  Anderson  Independent   6.6   loss
1984    Reagan   Republican  58.8    win
1984   Mondale   Democratic  37.6   loss
1988      Bush   Republican  53.4    win
1988   Dukakis   Democratic  45.6   loss
1992   Clinton   Democratic  43.0    win
1992      Bush   Republican  37.4   loss
1992     Perot  Independent  18.9   loss
1996   Clinton   Democratic  49.2    win
1996      Dole   Republican  40.7   loss
1996     Perot  Independent   8.4   loss
2000      Gore   Democratic  48.4   loss
2000      Bush   Republican  47.9    win
2004     Kerry   Democratic  48.3   loss
2004      Bush   Republican  50.7    win
2008     Obama   Democratic  52.9    win
2008    McCain   Republican  45.7   loss
2012     Obama   Democratic  51.1    win
2012    Romney   Republican  47.2   loss
2016   Cli

Year Candidate        Party     % Result
0   1980    Reagan   Republican  50.7    win
1   1980    Carter   Democratic  41.0   loss
2   1980  Anderson  Independent   6.6   loss
3   1984    Reagan   Republican  58.8    win
4   1984   Mondale   Democratic  37.6   loss
5   1988      Bush   Republican  53.4    win
6   1988   Dukakis   Democratic  45.6   loss
7   1992   Clinton   Democratic  43.0    win
8   1992      Bush   Republican  37.4   loss
9   1992     Perot  Independent  18.9   loss
10  1996   Clinton   Democratic  49.2    win
11  1996      Dole   Republican  40.7   loss
12  1996     Perot  Independent   8.4   loss
13  2000      Gore   Democratic  48.4   loss
14  2000      Bush   Republican  47.9    win
15  2004     Kerry   Democratic  48.3   loss
16  2004      Bush   Republican  50.7    win
17  2008     Obama   Democratic  52.9    win
18  2008    McCain   Republican  45.7   loss
19  2012     Obama   Democratic  51.1    win
20  2012    Romney   Republican  47.2   loss
21  2016   Clinton   Democratic  48.2   loss
22  2016     Trump   Republican  46.1    win

# Caution:
The **set_index command** (along with all other data frame methods) **does not modify the dataframe**, i.e., the original "elections" is untouched. Note: There is a flag called "inplace" which does modify the calling dataframe (e.g., `elections.set_index("Party",inplace=True)`).

## Duplicate Columns?
By contast, column names MUST be unique. For example, if we try to read in a file for which column names are not unique, Pandas will automatically any duplicates. Load duplicate_columns.csv

In [10]:
columnsShow = pd.read_csv('/workspaces/Advance_DS/duplicate_columns.csv')
print('Duplicate Column File read :\n' , columnsShow)
#       -----------------------------------------------------

ColumnShow_head = columnsShow.head()
print('First five rows show :\n' , ColumnShow_head)

Duplicate Column File read :
     name    name.1      flavor
0   john     smith     vanilla
1  zhang      shan   chocolate
2  fulan  alfulani  strawberry
3   hong   gildong      banana
First five rows show :
     name    name.1      flavor
0   john     smith     vanilla
1  zhang      shan   chocolate
2  fulan  alfulani  strawberry
3   hong   gildong      banana


## The [ ] Operator & Indexing

The DataFrame class has an indexing operator **[ ]** (also known as the 'brack' operator) that lets you do a variety of different things. If your provide a String to the **[ ]** operator, you get back a ***Series*** corresponding to the requested label.

1.Use **[ ]** to display different columns

2.Use List retrive multiple columns

In [11]:
# Answer Here
# 1.Use [ ] to display different columns

index_Operator_1 = columnsShow['name']
print('Only use Bracket show column name\n\n',index_Operator_1)
#             ---------------------------------------------
index_Operator_2 = columnsShow['name.1']
print('\nOnly use Bracket show column name.1\n\n',index_Operator_2)
#             ---------------------------------------------
index_Operator_3 = columnsShow['flavor']
print('\nOnly use Bracket show column flavor\n\n',index_Operator_3)

# 2.Use List retrive multiple columns
index_Operator = columnsShow[['name','flavor','name.1']]
print('\n\nUse List\n\n',index_Operator)

Only use Bracket show column name

 0     john
1    zhang
2    fulan
3     hong
Name: name, dtype: object

Only use Bracket show column name.1

 0       smith
1        shan
2    alfulani
3     gildong
Name: name.1, dtype: object

Only use Bracket show column flavor

 0       vanilla
1     chocolate
2    strawberry
3        banana
Name: flavor, dtype: object


Use List

     name      flavor    name.1
0   john     vanilla     smith
1  zhang   chocolate      shan
2  fulan  strawberry  alfulani
3   hong      banana   gildong


The **[ ]** operator also accepts a list of strings. In this case, you get back a **DataFrame** corresponding to the requested strings.

In [ ]:
# Answer Here
dataFrame1 = columnsShow[['name','flavor','name.1']]
print('\n\nUse List\n\n',dataFrame1)



Use List

     name      flavor    name.1
0   john     vanilla     smith
1  zhang   chocolate      shan
2  fulan  strawberry  alfulani
3   hong      banana   gildong


A list of one label also returns a DataFrame. This can be handy if you want your results as a DataFrame, not a series.

Note that we can also use the **to_frame** method to turn a Series into a DataFrame.

Extract one col name "Candidates" from DataFrame it will be a series. Convert series into a DataFrame.

In [ ]:
# Answer Here
print(columnsShow)
dupplicate_columns = columnsShow['name']
convert_duplicate_columns = dupplicate_columns.to_frame()

print('\n\nConvert Series into DataFrame\n\n',convert_duplicate_columns)

    name    name.1      flavor
0   john     smith     vanilla
1  zhang      shan   chocolate
2  fulan  alfulani  strawberry
3   hong   gildong      banana


Convert Series into DataFrame

     name
0   john
1  zhang
2  fulan
3   hong


### Row Indexing

The `[]` operator also accepts numerical slices as arguments. In this case, we are indexing by row, not column!

Extract few rows from DataFrame

In [ ]:
# Answer Here
extract_rows = dataFrame[1:10]
print(extract_rows)



                   Candidate                  Party  Popular vote Result  \
Year                                                                       
1824       John Quincy Adams  Democratic-Republican        113142    win   
1828          Andrew Jackson             Democratic        642806    win   
1828       John Quincy Adams    National Republican        500897   loss   
1832          Andrew Jackson             Democratic        702735    win   
1832              Henry Clay    National Republican        484205   loss   
1832            William Wirt           Anti-Masonic        100715   loss   
1836       Hugh Lawson White                   Whig        146109   loss   
1836        Martin Van Buren             Democratic        763291    win   
1836  William Henry Harrison                   Whig        550816   loss   

              %  
Year             
1824  42.789878  
1828  56.203927  
1828  43.796073  
1832  54.574789  
1832  37.603628  
1832   7.821583  
1836  10.005985  
1

If you provide a single argument to the `[]` operator, it tries to use it as a name. This is true even if the argument passed to **[ ]** is an integer.

In [ ]:
# dataFrame[0] #this does not work, try uncommenting this to see it fail in action, woo

The following cells allow you to **test your understanding**. Let's go over the summary of what we have learnt (see slides).

# Creating DataFrames
Create DataFrame using List and Columns name.

In [ ]:
# Answer Here

dataFrame_create = pd.DataFrame([['P1','M'],
                                 ['P2','F'],
                                 ['P3','F'],
                                 ['P4','M']],
                                columns=['name','Gender'])
print(dataFrame_create)

  name Gender
0   P1      M
1   P2      F
2   P3      F
3   P4      M


Creating DataFrames using **Dictionary**.

In [ ]:
# Answer Here
dataFrame_create_1 = pd.DataFrame({'name':['P1','P2','P3','P4'],
                                   'Gender':['M','F','F','M']})
print(dataFrame_create_1)

  name Gender
0   P1      M
1   P2      F
2   P3      F
3   P4      M


## Filtering via Boolean Array Selection

The `[]` operator also supports array of booleans as an input. In this case, the array must be exactly as long as the number of rows. The result is a **filtered version of the data frame**, where **only rows corresponding to True appear**.

In [ ]:
elections[[False, False, False, False, False,
          False, False, True, False, False,
          True, False, False, False, True,
          False, False, False, False, False,
          False, True, False]]

One very common task in Data Science is **filtering**. Boolean Array Selection is one way to achieve this in Pandas. We start by observing that **logical operators** like the equality operator can be applied to **Pandas Series data** to generate a **Boolean Array**.

Compare the 'Result' column to the String 'win' and Show results

In [ ]:
#Answer Here

Compare the 'Party' column to the String 'Democratic' and Show results

In [ ]:
#Answer Here

The output of the logical operator applied to the Series is **another Series with the same name and index, but of datatype boolean**.

These boolean Series can be used as an argument to the `[]` operator.

Creates  DataFrame of all election winners since 1980.

In [ ]:
#Answer Here

Above, we've assigned the result of the logical operator to a new variable called `iswin`. This is uncommon. Usually, the series is created and used on the same line. Such code is a little tricky to read at first, but you'll get used to it quickly.

Show all 'win' results between 1980 to 2000

In [ ]:
#Answer Here

Show all 'Loss' results of Independent party

In [ ]:
# Answer Here

We can select multiple criteria by creating multiple boolean Series and combining them using the `&` operator.

Show results of win with percentage less than 50%

In [ ]:
# Answer Here

Show all 'win' results between 1980 to 2000

In [ ]:
# Answer Here

## Loc and iLoc

Show 5 enteries from start

In [ ]:
# Answer Here

You can provide `.loc` a list of row labels [0-5] and column labels ['Candidate','Party', 'Year'] as input to return a dataframe

In [ ]:
#Answer Here

Loc also supports **slicing** (for all types, including numeric and string labels!). Note that the slicing for loc is **inclusive**, even for numeric slices.

Use Slicing on Rows and Columns

In [ ]:
# Answer Here

If we provide only a **single label** for the column argument, we get back a **Series**.

In [ ]:
# Answer Here

If we want a data frame instead and don't want to use to_frame, we can provide a **list** containing the column name.

In [ ]:
# Answer Here

If we give only one row but many column labels, we'll get back a **Series** corresponding to a row of the table. This new Series has a neat index, where **each entry is the name of the column** that the data came from.

In [ ]:
# Answer Here

In [ ]:
# Answer Here

If we omit the column argument altogether, the **default behavior is to retrieve all columns**.

In [ ]:
# Answer Here


Specify Rows and Columns as List to retrive specific enteries

In [ ]:
# Answer Here

Boolean Series are also boolean arrays, so we can use the Boolean Array Selection from earlier using loc as well.

In [ ]:
# Answer Here

## String-labeled Rows

Let's do a quick example using data with string-labeled rows instead of integer labeled rows, just to make sure we're really understanding loc.

Use mottos.csv file

In [ ]:
# Answer Here

Extract slice, can be specified using slice notation, even if the rows have string labels instead of integer labels.

### iloc

loc's cousin iloc is very similar, but is used to access based on numerical position instead of label. For example, to access to the top 3 rows and top 3 columns of a table, we can use [0:3, 0:3]. 'iloc' slicing is **exclusive**, just like standard Python slicing of numerical values.

Use iloc to extract first 3 rows and columns from elections DataFrame

In [ ]:
#Answer Here

We will use both `loc` and `iloc` in the course. `loc` is generally preferred for a number of reasons, for example:

1. It is harder to make mistakes since you have to literally write out what you want to get.
2. Code is easier to read, because the reader doesn't have to know e.g., what column #17 represents.
3. It is robust against permutations of the data, e.g. the social security administration switches the order of two columns.

However, iloc is sometimes more convenient. We'll provide examples of when iloc is the superior choice.

## Handy Properties and Utility Functions for Series and DataFrames

The head, shape, size, and describe methods can be used to quickly get a good sense of the data we're working with. For example:

In [13]:
mottos = pd.read_csv("/workspaces/Advance_DS/mottos.csv")

In [14]:
# Answer Here

print('Print first 5 rows using head command : \n', mottos.head())
print('Print first 5 rows using tail command : ', mottos.tail())
print('Print first 5 rows using describe command : ', mottos.describe())

Print first 5 rows using head command : 
         State                          Motto                 Translation  \
0     Alabama  Audemus jura nostra defendere  We dare defend our rights!   
1      Alaska            North to the future                           —   
2     Arizona                     Ditat Deus                God enriches   
3    Arkansas                 Regnat populus             The people rule   
4  California                Eureka (Εὕρηκα)             I have found it   

  Language Date Adopted  
0    Latin         1923  
1  English         1967  
2    Latin         1863  
3    Latin         1907  
4    Greek         1849  
Print first 5 rows using tail command :              State                  Motto                   Translation  \
45       Virginia    Sic semper tyrannis        Thus always to tyrants   
46     Washington          Al-ki or Alki                     By and by   
47  West Virginia  Montani semper liberi  Mountaineers are always free   
48      

In [15]:
# Answer Here

print('Print first 5 rows using head command \t: \n', mottos.head())
print('\nPrint first 5 rows using tail command \t: ', mottos.tail())
print('\nPrint first 5 rows using describe command : ', mottos.describe())

Print first 5 rows using head command 	: 
         State                          Motto                 Translation  \
0     Alabama  Audemus jura nostra defendere  We dare defend our rights!   
1      Alaska            North to the future                           —   
2     Arizona                     Ditat Deus                God enriches   
3    Arkansas                 Regnat populus             The people rule   
4  California                Eureka (Εὕρηκα)             I have found it   

  Language Date Adopted  
0    Latin         1923  
1  English         1967  
2    Latin         1863  
3    Latin         1907  
4    Greek         1849  

Print first 5 rows using tail command 	:              State                  Motto                   Translation  \
45       Virginia    Sic semper tyrannis        Thus always to tyrants   
46     Washington          Al-ki or Alki                     By and by   
47  West Virginia  Montani semper liberi  Mountaineers are always free   
48   

Size of DataFrame

In [ ]:
# Answer Here
print('Size of Data Frame : ', mottos.size)

Size of Data Frame :  250


The fact that the size is 250 means our data file is relatively small, with only 250 total entries.

Shape of DataFrame

In [ ]:
# Answer Here

Use describe function and extract the meaningful information from DataFrame

In [ ]:
# Answer Here

Above, we see a quick summary of all the data. For example, the most common language for mottos is Latin, which covers 23 different states. Does anything else seem surprising?

We can get a direct reference to the index using .index.

In [ ]:
# Answer Here

In [ ]:
mottos.head(2)

It turns out the columns also have an Index. We can access this index by using `.columns`.

In [ ]:
# Answer Here

Index(['State', 'Motto', 'Translation', 'Language', 'Date Adopted'], dtype='object')

## Sorting and Value Counts

There are also a ton of useful utility methods we can use with Data Frames and Series. For example, we can create a copy of a data frame sorted by a specific column using `sort_values`.

In [ ]:
# Answer Here

As mentioned before, all Data Frame methods return a copy and do **not** modify the original data structure, unless you set inplace to True.

If we want to sort in reverse order, we can set `ascending=False`.

In [ ]:
elections.sort_values('%', ascending=False)

We can also use `sort_values` on Series objects.

In [ ]:
mottos['Language'].sort_values().head(50)

For Series, the `value_counts` method is often quite handy.

In [ ]:
mottos['Language'].value_counts()

Also commonly used is the `unique` method, which returns **all unique values** as a numpy array.

In [ ]:
mottos['Language'].unique()

In [ ]:
def fiba(n):
    if n < 2:
        return n
    else:
        return fiba(n-1) + fiba(n-2)



fiba(5)

5

# Thank you!